itests/kit/node_unmanaged.go

```diff
@@ -464,332 +685,424 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
+		if exitCodes[i] == exitcode.Ok {
+			req.NoError(err)
+			req.Equal(si.SectorNumber, sector.SectorNumber)
+			req.Equal(si.Activation, msgReturn.Height-1)
+		} else {
+			req.Nil(si, "sector should not be on chain")
+		}
+		if proofType.IsNonInteractive() {
+			// NI-PoRep lets us determine the deadline, so we can check that it's set correctly
+			sp, err := tm.FullNode.StateSectorPartition(tm.ctx, tm.ActorAddr, sector.SectorNumber, msgReturn.TipSet)
+			if exitCodes[i] == exitcode.Ok {
+				req.NoError(err)
+				req.Equal(provingDeadline, sp.Deadline)
+			} else {
+				req.ErrorContains(err, fmt.Sprintf("sector %d not due at any deadline", sector.SectorNumber))
+			}
+		}
 	}
 
-	// Calculate the exact epoch when proving should occur
-	provingEpoch := periodStart + di.WPoStChallengeWindow*abi.ChainEpoch(sp.Deadline)
-
-	tm.t.Logf("Miner %s: WindowPoST(%d): next ProvingEpoch: %d", tm.ActorAddr, sectorNumber, provingEpoch)
-
-	return di.CurrentEpoch, provingEpoch, nil
+	return exitCodes
 }
 
-func (tm *TestUnmanagedMiner) wdPostLoop(
-	pctx context.Context,
-	sectorNumber abi.SectorNumber,
-	sealedCid cid.Cid,
-	sealedPath,
-	cacheDir string,
-) (chan WindowPostResp, context.CancelFunc) {
-
-	ctx, cancelFn := context.WithCancel(pctx)
-	respCh := make(chan WindowPostResp, 1)
+func (tm *TestUnmanagedMiner) wdPostLoop() {
+	if tm.runningWdPostLoop {
+		return
+	}
+	tm.runningWdPostLoop = true
 
-	head, err := tm.FullNode.ChainHead(ctx)
+	head, err := tm.FullNode.ChainHead(tm.ctx)
 	require.NoError(tm.t, err)
 
 	// wait one challenge window for cron to do its thing with deadlines, just to be sure so we get
 	// an accurate dline.Info whenever we ask for it
-	_ = tm.FullNode.WaitTillChain(ctx, HeightAtLeast(head.Height()+miner14.WPoStChallengeWindow+5))
+	_ = tm.FullNode.WaitTillChain(tm.ctx, HeightAtLeast(head.Height()+miner14.WPoStChallengeWindow+5))
 
 	go func() {
-		var firstPost bool
+		var postCount int
 
-		writeRespF := func(respErr error) {
-			var send WindowPostResp
-			if respErr == nil {
-				if firstPost {
-					return // already reported on our first post, no error to report, don't send anything
-				}
-				send.Posted = true
-				firstPost = true
-			} else {
-				if ctx.Err() == nil {
-					tm.t.Logf("Sector %d: WindowPoSt submission failed: %s", sectorNumber, respErr)
-				}
-				send.Error = respErr
+		recordPostOrError := func(post windowPost) {
+			head, err := tm.FullNode.ChainHead(tm.ctx)
+			if err == nil {
+				post.Epoch = head.Height()
 			}
-			select {
-			case respCh <- send:
-			case <-ctx.Done():
-			default:
+			if post.Error != nil && tm.ctx.Err() == nil {
+				tm.log("WindowPoSt submission failed for sectors %v at epoch %d: %s", post.Posted, post.Epoch, post.Error)
+			} else if tm.ctx.Err() == nil {
+				postCount++
+				tm.log("WindowPoSt loop completed post %d at epoch %d for sectors %v", postCount, post.Epoch, post.Posted)
 			}
+			tm.postsLk.Lock()
+			tm.posts = append(tm.posts, post)
```

Comment:
We should not do this if ` tm.ctx.Err()` is not nil.

Otherwise, there is a race where this is called after the context is cancelled(post submission errors out because of the context cancellation) but we wrongly mark the post as successful by appending it here and error is nil in the appended post.